JSON serialiser now calculates hash over schema

BernhardBln opened the pull request.

uweschaefer commented on factcast-core/pom.xml:
```diff
@@ -1,64 +1,71 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
what is the change here?

BernhardBln replied:
No clue, probably from spotless plugin?

uweschaefer commented on factcast-core/src/main/java/org/factcast/core/snap/SnapshotId.java:
```diff
@@ -29,4 +29,5 @@
     @NonNull
     UUID uuid; // semantically optional, might be constant if not needed
                // (SnapshotProjections)
+
```
:grin:

BernhardBln replied:
das sah aber auch aus :D

uweschaefer commented on factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:
```diff
@@ -158,10 +177,21 @@ public static JsonNode toJsonNode(Map<String, Object> jsonAsMap) {
     }
+
+    @SneakyThrows
+    public static long calculateHash(Class<?> projectionClass) {
```
this uses the FCJson internal objectmapper, but we agreed that people must be able to swap the obejctmapper used for eventserialization. therefore i'd suggest to move this out of here and into the eventserializer ?

uweschaefer commented on factcast-factus-bin-snapser/pom.xml:
```diff
@@ -1,46 +1,53 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
what is the change here?

BernhardBln replied:
added json-schema dependency

uweschaefer commented on factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:
```diff
@@ -158,10 +177,21 @@ public static JsonNode toJsonNode(Map<String, Object> jsonAsMap) {
+    @SneakyThrows
+    public static long calculateHash(Class<?> projectionClass) {
+
+        JsonSchema jsonSchema = schemaGen.generateSchema(projectionClass);
```
also i guess, the schema should be cached (Map<Type,Schema>)

uweschaefer replied:
or better Ma<Type,Long> directly to the md5

BernhardBln replied:
Done in AbstractSnapshotRepository

uweschaefer commented on factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:
```diff
@@ -158,10 +177,21 @@ public static JsonNode toJsonNode(Map<String, Object> jsonAsMap) {
+                .writeValueAsString(jsonSchema);
+
+        return schemaModifier.apply(schema)
+                .hashCode();
```
is the hashcode even guaranteed to be consistent? i'd expect taking the md5 of the string

BernhardBln replied:
changed

uweschaefer reviewed (changes requested):
i'd suggest 

* MD5 instead of schema hash
* not recalculating the hash for a given class

BernhardBln commented on factcast-core/pom.xml:
```diff
@@ -1,64 +1,71 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-jsonSchema</artifactId>
+            <version>2.11.2</version>
```
How does it work with renovate-bot, would we remove the version here? Also above, in databind?

uweschaefer replied:
i think there is a version defined in the properties. "jackson.version"?

BernhardBln commented:
@uweschaefer I cleaned up dependency management for jackson a bit:

- introduced property with version in root pom
- used this property for all jackson dependencies in bom
- removed version tag in all other poms

BernhardBln commented:
@uweschaefer what does that mean?

```
Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.4.1:check (default) on project factcast: 
     Execution default of goal com.diffplug.spotless:spotless-maven-plugin:2.4.1:check failed: 
          No such reference 'origin/master' -
```

https://github.com/factcast/factcast/pull/1050/checks?check_run_id=1175889328

uweschaefer commented:
crap. spotless is configured to change only files diffing compared to master. the message implies, that the repo is not fully cloned, but rather just a particular branch is sparse checked out. will look into it. thx

uweschaefer commented:
thx @BernhardBln
